ostd/src/task/scheduler/mod.rs

```diff
@@ -0,0 +1,246 @@
+pub fn inject_scheduler(scheduler: &'static dyn Scheduler) {
+    SCHEDULER.call_once(|| scheduler);
+
+    register_callback(|| {
+        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |local_rq| {
+            if local_rq.update_current(UpdateFlags::Tick) {
+                local_rq.set_should_preempt(true);
+            }
+        })
+    });
+}
+
+/// Adds task to scheduler.
+///
+/// This function may be called with newly spawned task or waked task.
+pub fn add_task(runnable: Arc<Task>, flags: EnqueueFlags) {
+    if !SCHEDULER.is_completed() {
+        let fifo_scheduler = Box::new(FifoScheduler::<Task>::new());
+        let static_scheduler = Box::leak(fifo_scheduler);
+        inject_scheduler(static_scheduler);
+    }
+    SCHEDULER.get().unwrap().enqueue(runnable, flags);
+}
+
+/// Yields execution.
+pub fn yield_now(flags: YieldFlags) {
+    reschedule(&mut |local_rq| {
+        if flags != YieldFlags::Exit {
+            local_rq.update_current(flags.into());
+        }
+
+        let mut current_task_option = None;
+        if flags != YieldFlags::Yield {
+            current_task_option = local_rq.dequeue_current();
+        }
+        let current_task;
+        let mut current_task_inner = None;
+        if flags == YieldFlags::Wait {
+            current_task = current_task_option.clone().unwrap();
+            current_task_inner = Some(current_task.inner_exclusive_access());
+
+            if current_task_inner.as_ref().unwrap().task_status == TaskStatus::Runnable {
+                drop(current_task_inner);
+                local_rq.set_current(current_task_option);
+                return ReschedAction::DoNothing;
+            }
+        }
+
+        if let Some(next) = local_rq.pick_next_current() {
+            if flags == YieldFlags::Wait {
+                debug_assert_eq!(
+                    current_task_inner.as_ref().unwrap().task_status,
+                    TaskStatus::Sleepy
+                );
+                current_task_inner.unwrap().task_status = TaskStatus::Sleeping;
+            }
+            local_rq.set_current(Some(next.clone()));
+            ReschedAction::SwitchTo(next)
+        } else {
+            match flags {
+                YieldFlags::Exit => ReschedAction::Retry,
+                YieldFlags::Yield => ReschedAction::DoNothing,
+                YieldFlags::Wait => {
+                    drop(current_task_inner);
+                    local_rq.set_current(current_task_option);
+                    ReschedAction::DoNothing
+                }
+            }
+        }
+    })
+}
+
+/// Invokes a schedule.
+pub fn schedule() {
+    reschedule(&mut |local_rq| {
+        if !local_rq.should_preempt() {
+            return ReschedAction::DoNothing;
+        }
+        local_rq.set_should_preempt(false);
+
+        if let Some(next_current) = local_rq.pick_next_current() {
+            local_rq.set_current(Some(next_current.clone()));
```

Comment:
Here is the logic of sleeping separated from the current messy `yield_now`.
```rust
pub fn sleep() {
    reschedule(&mut |local_rq| {
        local_rq.update_current(flags.into());
        let mut current_task = local_rq.dequeue_current().unwrap();
        let mut current_task_inner = current_task.inner_exclusive_access();

        if current_task_inner.task_status == TaskStatus::Runnable {
            // An irq wake up the task before we `pick_next_current`.
            drop(current_task_inner);

            local_rq.set_current(Some(current_task));
            return ReschedAction::DoNothing;
        }

        if let Some(next) = local_rq.pick_next_current() {
            debug_assert_eq!(
                current_task_inner.task_status,
                TaskStatus::Sleepy
            );
            current_task_inner.task_status = TaskStatus::Sleeping;
            drop(current_task_inner);

            local_rq.set_current(Some(next.clone()));
            ReschedAction::SwitchTo(next)
        } else {
            // There is no qualified candidate for `next`.
            drop(current_task_inner);

            local_rq.set_current(Some(current_task);
            ReschedAction::DoNothing
        }
    })
}
```
The `set_current` is for the cases when the current is waken up before we pick next current and when we cannot find a qualified candidate for next current. Merging the logic of `pick_next_current` and `set_current` can make life much harder in these two cases.

As to the invariance, it should be OK if scheduler developers follow the contract specified by the comment below since we (ostd) are the only user and we ensure it's safe when used. 
```rust
    /// Sets the current task with given `next`.
    ///
    /// The runqueue should keep the previous current in.
    fn set_current(&mut self, next: Option<Arc<T>>);
```
However, I do think leaving part of this to the scheduler developers (i.e, policy makers) is (maybe) kind of irresponsible. However, I didn't work out a better solution. Could you please give some advice on this?